### PR TITLE
functions: add Replace, and RegexpReplace from Terraform

### PR DESCRIPTION
**Pull request by azr**

### CODE REVIEW

**Comment by azr** [comment]
Hey @apparentlymart ! No worries, thanks for reviewing, I had kept the names of the go standard library, because, naming is hard !

I agree with you that from a HCL/CTY standing point it makes more sense to have a replace call replace everything. In programming people sorts of often have the performances in mind.

Changing this.

**Comment by azr** [comment]
Here we go, I also added some tests !

File: cty/function/stdlib/string_replace.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	},
+	Type: function.StaticReturnType(cty.String),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+		str := args[0].AsString()
+		substr := args[1].AsString()
+		replace := args[2].AsString()
+
+		// We search/replace using a regexp if the string is surrounded
+		// in forward slashes.
+		if len(substr) > 1 && substr[0] == '/' && substr[len(substr)-1] == '/' {
```
Comment (apparentlymart) [inline cty/function/stdlib/string_replace.go]: (I'm sorry I didn't catch this on the initial read yesterday)

Now that we have a separate function name as a more "normal" way to select between plain string or regex-based replacement, I'd prefer to drop this special case and just expect the whole string to be a valid regular expression. That will then be consistent with the behavior of the existing `Regex` and `RegexAll` functions already in this package. I think if leading and trailing slashes are present then they should be interpreted as a literal part of the given pattern, not as special markers.

The rest of this looks great! Thanks for the updates.

Comment (azr) [reply cty/function/stdlib/string_replace.go]: Awesome !

File: cty/function/stdlib/string_replace.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+			substr = substr[1 : len(substr)-1]
+		}
```
Comment (azr) [inline cty/function/stdlib/string_replace.go]: [test] May be I don't even have to open an editor ? : 

```suggestion
```

Comment (azr) [reply cty/function/stdlib/string_replace.go]: Nice !

File: cty/function/stdlib/string_replace.go (hunk 1 above)
Comment (azr) [reply cty/function/stdlib/string_replace.go]: Great thanks for reviewing I think this is good to go 🙂

File: cty/function/stdlib/string_replace.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+			Name: "replace",
+			Type: cty.String,
+		},
+	},
+	Type: function.StaticReturnType(cty.String),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+		str := args[0].AsString()
+		substr := args[1].AsString()
+		replace := args[2].AsString()
+
```
Comment (azr) [inline cty/function/stdlib/string_replace.go]:
```suggestion
```

File: cty/function/stdlib/string_replace_test.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+func TestRegexReplace(t *testing.T) {
+	tests := []struct {
+		Input           cty.Value
+		Substr, Replace cty.Value
+		Want            cty.Value
+	}{
+		{
+			cty.StringVal("-ab-axxb-"),
+			cty.StringVal("/a(x*)b/"),
```
Comment (azr) [inline cty/function/stdlib/string_replace_test.go]:
```suggestion
			cty.StringVal("a(x*)b"),
```

**Comment by apparentlymart** [comment]
Thanks! :tada:

**Comment by apparentlymart** [comment]
While I was writing the changelog entry for this I noticed that the new function was called `RegexpReplace` while the existing similar functions were called `Regex` and `RegexAll` (without the "p").

For consistency, I made follow up commit 440c597d573e4b979f577dd3d0f8a23aee2b318a to rename to `RegexReplace`. The implementation is otherwise unchanged. Sorry I didn't catch that before!

**Comment by azr** [comment]
Ah no worries :) Thanks !!! 🚀 🚀 🚀

**Pull request merged**